### ISSUE TITLE
[PATCH] Upgrade to threads 1.78

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#10656]
Migrated from [rt.perl.org#78000](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=78000) (status was 'resolved')

Searchable as RT78000$

**Comment by p5pRT** [comment]
### From @jdhedden

#### Created by @jdhedden



The attached patch upgrades the 'threads' module to v1\.78\.

This upgrade fixes 'threads' bug \#60460
&nbsp;    https://rt.cpan.org/Ticket/Display.html?id=60460
which results in segfaults if a signal is sent to a thread\, and
there is no corresponding signal handler set in the thread\.
Corresponding tests are included\.

Additionally\, the POD is updated to document \[perl \#75154\] concerning
open dir handles and threads\.

The module version for blead is set to 1\.78\_01 to account for
blead changes not present in the CPAN version\.

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=medium
    module=threads

Site configuration information for perl 5.12.2:

Configured by Jerry at Thu Sep 16 09:52:43 EDT 2010.

Summary of my perl5 (revision 5 version 12 subversion 2 patch 42531)
configuration:
  Snapshot of: 7a3b65c9d99f69553fffe01f73d49fe9abf95120
  Platform:
    osname=cygwin, osvers=1.5.25(0.15642), archname=cygwin-thread-multi-64int
    uname='cygwin_nt-5.1 pn100-02-2-364p 1.5.25(0.15642) 2008-06-12
19:34 i686 cygwin '
    config_args='-de -Duse64bitint -Dusethreads
-Dinc_version_list=none -Dnoextensions=IPC/SysV Sys/Syslog Devel/DProf
Devel/Peek re GDBM_File NDBM_File ODBM_File Text/Soundex
Math/BigInt/FastCalc Time/Piece -A append:ccflags= -DNO_MATHOMS  -A
define:optimize=-Os -pipe -funit-at-a-time -march=pentium4
-mfpmath=sse -mieee-fp -mmmx -msse -msse2'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=undef, uselongdouble=undef
    usemymalloc=y, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-DPERL_USE_SAFE_PUTENV -U__STRICT_ANSI__
-DNO_MATHOMS -fno-strict-aliasing -pipe',
    optimize='-Os -pipe -funit-at-a-time -march=pentium4 -mfpmath=sse
-mieee-fp -mmmx -msse -msse2',
    cppflags='-DPERL_USE_SAFE_PUTENV -U__STRICT_ANSI__ -DNO_MATHOMS
-fno-strict-aliasing -pipe'
    ccversion='', gccversion='3.4.4 (cygming special, gdc 0.12, using
dmd 0.125)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long long', ivsize=8, nvtype='double', nvsize=8,
Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='g++', ldflags =' -Wl,--enable-auto-import
-Wl,--export-all-symbols -Wl,--stack,8388608
-Wl,--enable-auto-image-base -s -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib /lib
    libs=-lgdbm -ldl -lcrypt -lgdbm_compat
    perllibs=-ldl -lcrypt
    libc=/usr/lib/libc.a, so=dll, useshrplib=true, libperl=cygperl5_12_2.dll
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' -s'
    cccdlflags=' ', lddlflags=' --shared  -Wl,--enable-auto-import
-Wl,--export-all-symbols -Wl,--stack,8388608
-Wl,--enable-auto-image-base -s -L/usr/local/lib'

Locally applied patches:



@INC for perl 5.12.2:
    /usr/lib/perl5/site_perl/5.12.2/cygwin
    /usr/lib/perl5/site_perl/5.12.2
    /usr/lib/perl5/5.12.2/cygwin
    /usr/lib/perl5/5.12.2
    .


Environment for perl 5.12.2:
    CYGWIN=ntsec
    HOME=/home/jhedden
    LANG=C
    LANGUAGE=C
    LC_ALL=C
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/jhedden/bin:/usr/local/src/perl/bin:/link/work/bin:/usr/local/bin:/usr/bin:/bin:/usr/X11R6/bin:/c/Program
Files/WiX:/c/djgpp/bin:/c/bb/tools/apache-ant-1.6.5/bin:/c/Program
Files/nant-0.85/bin:/c/bb/tools/jse5/bin:/c/dev-cpp/bin:/c/WINDOWS/system32:/c/WINDOWS:/c/WINDOWS/system32/WBEM:/c/blp/API/dde:/c/blp/API:/c/oracle/ora92/bin:/c/Program
Files/Oracle/jre/1.3.1/bin:/c/Program
Files/Oracle/jre/1.1.8/bin:/c/Program
Files/Hummingbird/Connectivity/7.10/Accessories:/usr/bin:/c/Program
Files/Windows Imaging/:/c/Program Files/Citrix/ICAService/:/c/Program
Files/Citrix/System32/:.
    PERLIO=perlio
    PERL_BADLANG (unset)
    SHELL (unset)

```
</details>

**Comment by p5pRT** [comment]
### From @jdhedden

<details><summary>0001-Upgrade-to-threads-1.78.patch</summary>

```diff
From fd07df67bc9b4c1f5bda200a68389795b8bb40f3 Mon Sep 17 00:00:00 2001
From: Jerry D. Hedden <jdhedden@cpan.org>
Date: Wed, 22 Sep 2010 13:28:48 -0400
Subject: [PATCH] Upgrade to threads 1.78

---
 MANIFEST                    |    1 +
 Porting/Maintainers.pl      |    2 +-
 dist/threads/lib/threads.pm |   12 ++++++--
 dist/threads/t/exit.t       |   10 +++---
 dist/threads/t/kill2.t      |   68 +++++++++++++++++++++++++++++++++++++++++++
 dist/threads/t/thread.t     |    2 +-
 dist/threads/threads.xs     |   48 ++++++++++++++++++++----------
 7 files changed, 117 insertions(+), 26 deletions(-)
 create mode 100644 dist/threads/t/kill2.t

diff --git a/MANIFEST b/MANIFEST
index 3e9583a..e132c72 100644
--- a/MANIFEST
+++ b/MANIFEST
@@ -2973,6 +2973,7 @@ dist/threads/t/free2.t		More ithread destruction tests
 dist/threads/t/free.t		Test ithread destruction
 dist/threads/threads.xs		ithreads
 dist/threads/t/join.t		Testing the join function
+dist/threads/t/kill2.t		Tests thread signalling
 dist/threads/t/kill.t		Tests thread signalling
 dist/threads/t/libc.t		testing libc functions for threadsafety
 dist/threads/t/list.t		Test threads->list()
diff --git a/Porting/Maintainers.pl b/Porting/Maintainers.pl
index 4b153bb..fabc35a 100755
--- a/Porting/Maintainers.pl
+++ b/Porting/Maintainers.pl
@@ -1436,7 +1436,7 @@ use File::Glob qw(:case);
     'threads' =>
 	{
 	'MAINTAINER'	=> 'jdhedden',
-	'DISTRIBUTION'	=> 'JDHEDDEN/threads-1.77.tar.gz',
+	'DISTRIBUTION'	=> 'JDHEDDEN/threads-1.78.tar.gz',
 	'FILES'		=> q[dist/threads],
 	'EXCLUDED'	=> [ qr{^examples/},
 			     qw(t/pod.t
diff --git a/dist/threads/lib/threads.pm b/dist/threads/lib/threads.pm
index 8654f9f..a1dd2b2 100644
--- a/dist/threads/lib/threads.pm
+++ b/dist/threads/lib/threads.pm
@@ -5,7 +5,7 @@ use 5.008;
 use strict;
 use warnings;
 
-our $VERSION = '1.77_03';
+our $VERSION = '1.78_01';
 my $XS_VERSION = $VERSION;
 $VERSION = eval $VERSION;
 
@@ -134,7 +134,7 @@ threads - Perl interpreter-based threads
 
 =head1 VERSION
 
-This document describes threads version 1.77
+This document describes threads version 1.78
 
 =head1 SYNOPSIS
 
@@ -1005,6 +1005,12 @@ mutexes that are needed to control functionality within the L<threads> module.
 For this reason, the use of C<END> blocks in threads is B<strongly>
 discouraged.
 
+=item Open directory handles
+
+Spawning threads with open directory handles (see
+L<opendir|perlfunc/"opendir DIRHANDLE,EXPR">) will crash the interpreter.
+L<[perl #75154]|http://rt.perl.org/rt3/Public/Bug/Display.html?id=75154>
+
 =item Perl Bugs and the CPAN Version of L<threads>
 
 Support for threads extends beyond the code in this module (i.e.,
@@ -1034,7 +1040,7 @@ L<threads> Discussion Forum on CPAN:
 L<http://www.cpanforum.com/dist/threads>
 
 Annotated POD for L<threads>:
-L<http://annocpan.org/~JDHEDDEN/threads-1.77/threads.pm>
+L<http://annocpan.org/~JDHEDDEN/threads-1.78/threads.pm>
 
 Source repository:
 L<http://code.google.com/p/threads-shared/>
diff --git a/dist/threads/t/exit.t b/dist/threads/t/exit.t
index 29c3dca..c63ca1c 100644
--- a/dist/threads/t/exit.t
+++ b/dist/threads/t/exit.t
@@ -48,7 +48,7 @@ my $rc = $thr->join();
 ok(! defined($rc), 'Exited: threads->exit()');
 
 
-run_perl(prog => 'use threads 1.77;' .
+run_perl(prog => 'use threads 1.78;' .
                  'threads->exit(86);' .
                  'exit(99);',
          nolib => ($ENV{PERL_CORE}) ? 0 : 1,
@@ -98,7 +98,7 @@ $rc = $thr->join();
 ok(! defined($rc), 'Exited: $thr->set_thread_exit_only');
 
 
-run_perl(prog => 'use threads 1.77 qw(exit thread_only);' .
+run_perl(prog => 'use threads 1.78 qw(exit thread_only);' .
                  'threads->create(sub { exit(99); })->join();' .
                  'exit(86);',
          nolib => ($ENV{PERL_CORE}) ? 0 : 1,
@@ -108,7 +108,7 @@ run_perl(prog => 'use threads 1.77 qw(exit thread_only);' .
     is($?>>8, 86, "'use threads 'exit' => 'thread_only'");
 }
 
-my $out = run_perl(prog => 'use threads 1.77;' .
+my $out = run_perl(prog => 'use threads 1.78;' .
                            'threads->create(sub {' .
                            '    exit(99);' .
                            '});' .
@@ -124,7 +124,7 @@ my $out = run_perl(prog => 'use threads 1.77;' .
 like($out, '1 finished and unjoined', "exit(status) in thread");
 
 
-$out = run_perl(prog => 'use threads 1.77 qw(exit thread_only);' .
+$out = run_perl(prog => 'use threads 1.78 qw(exit thread_only);' .
                         'threads->create(sub {' .
                         '   threads->set_thread_exit_only(0);' .
                         '   exit(99);' .
@@ -141,7 +141,7 @@ $out = run_perl(prog => 'use threads 1.77 qw(exit thread_only);' .
 like($out, '1 finished and unjoined', "set_thread_exit_only(0)");
 
 
-run_perl(prog => 'use threads 1.77;' .
+run_perl(prog => 'use threads 1.78;' .
                  'threads->create(sub {' .
                  '   $SIG{__WARN__} = sub { exit(99); };' .
                  '   die();' .
diff --git a/dist/threads/t/kill2.t b/dist/threads/t/kill2.t
new file mode 100644
index 0000000..a68e7bc
--- /dev/null
+++ b/dist/threads/t/kill2.t
@@ -0,0 +1,68 @@
+use strict;
+use warnings;
+
+BEGIN {
+    require($ENV{PERL_CORE} ? '../../t/test.pl' : './t/test.pl');
+
+    use Config;
+    if (! $Config{'useithreads'}) {
+        skip_all(q/Perl not compiled with 'useithreads'/);
+    }
+}
+
+use ExtUtils::testlib;
+
+use threads;
+
+BEGIN {
+    $| = 1;
+    print("1..3\n");   ### Number of tests that will be run ###
+};
+
+fresh_perl_is(<<'EOI', 'ok', { }, 'No signal handler in thread');
+    use threads;
+    my $test = sub {
+        while(1) { sleep(1) }
+    };
+    my $thr = threads->create($test);
+    threads->yield();
+    $thr->detach();
+    eval {
+        $thr->kill('STOP');
+    };
+    print(($@ =~ /no signal handler set/) ? 'ok' : 'not ok');
+EOI
+
+fresh_perl_is(<<'EOI', 'ok', { }, 'Handler to signal mismatch');
+    use threads;
+    my $test = sub {
+        $SIG{'TERM'} = sub { threads->exit() };
+        while(1) { sleep(1) }
+    };
+    my $thr = threads->create($test);
+    threads->yield();
+    $thr->detach();
+    eval {
+        $thr->kill('STOP');
+    };
+    print(($@ =~ /no signal handler set/) ? 'ok' : 'not ok');
+EOI
+
+fresh_perl_is(<<'EOI', 'ok', { }, 'Handler and signal match');
+    use threads;
+    my $test = sub {
+        $SIG{'STOP'} = sub { threads->exit() };
+        while(1) { sleep(1) }
+    };
+    my $thr = threads->create($test);
+    threads->yield();
+    $thr->detach();
+    eval {
+        $thr->kill('STOP');
+    };
+    print((! $@) ? 'ok' : 'not ok');
+EOI
+
+exit(0);
+
+# EOF
diff --git a/dist/threads/t/thread.t b/dist/threads/t/thread.t
index b390215..b46a526 100644
--- a/dist/threads/t/thread.t
+++ b/dist/threads/t/thread.t
@@ -161,7 +161,7 @@ package main;
 
 # bugid #24165
 
-run_perl(prog => 'use threads 1.77;' .
+run_perl(prog => 'use threads 1.78;' .
                  'sub a{threads->create(shift)} $t = a sub{};' .
                  '$t->tid; $t->join; $t->tid',
          nolib => ($ENV{PERL_CORE}) ? 0 : 1,
diff --git a/dist/threads/threads.xs b/dist/threads/threads.xs
index f4e6946..d5e20b9 100644
--- a/dist/threads/threads.xs
+++ b/dist/threads/threads.xs
@@ -1342,6 +1345,7 @@ ithread_kill(...)
         ithread *thread;
         char *sig_name;
         IV signal;
+        int no_handler = 1;
     CODE:
         /* Must have safe signals */
         if (PL_signals & PERL_SIGNALS_UNSAFE_FLAG) {
@@ -1371,11 +1375,21 @@ ithread_kill(...)
         MUTEX_LOCK(&thread->mutex);
         if (thread->interp) {
             dTHXa(thread->interp);
-            PL_psig_pend[signal]++;
-            PL_sig_pending = 1;
+            if (PL_psig_pend && PL_psig_ptr[signal]) {
+                PL_psig_pend[signal]++;
+                PL_sig_pending = 1;
+                no_handler = 0;
+            }
+        } else {
+            /* Ignore signal to terminated thread */
+            no_handler = 0;
         }
         MUTEX_UNLOCK(&thread->mutex);
 
+        if (no_handler) {
+            Perl_croak(aTHX_ "Signal %s received in thread %"UVuf", but no signal handler set.", sig_name, thread->tid);
+        }
+
         /* Return the thread to allow for method chaining */
         ST(0) = ST(0);
         /* XSRETURN(1); - implied */
-- 
1.6.1.2


```
</details>

**Comment by p5pRT** [comment]
### From @jdhedden

Attached is a revised patch to upgrade to v1\.79\.  I needed to fix
the tests in the newly added t/kill2\.t file\.  The blead version will
be 1\.79\_01\.

On Wed\, Sep 22\, 2010 at 16&#8203;:08\, Jerry D\. Hedden
\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> \# New Ticket Created by  "Jerry D\. Hedden"
> \# Please include the string&#8203;:  \[perl \#78000\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=78000 >
>
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> The attached patch upgrades the 'threads' module to v1\.78\.
>
> This upgrade fixes 'threads' bug \#60460
>    https://rt.cpan.org/Ticket/Display.html?id=60460
> which results in segfaults if a signal is sent to a thread\, and
> there is no corresponding signal handler set in the thread\.
> Corresponding tests are included\.
>
> Additionally\, the POD is updated to document \[perl \#75154\] concerning
> open dir handles and threads\.
>
> The module version for blead is set to 1\.78\_01 to account for
> blead changes not present in the CPAN version\.
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

**Comment by p5pRT** [comment]
### From @jdhedden

<details><summary>0001-Upgrade-to-threads-1.79.patch</summary>

```diff
From 2c703e8ae49348185b499747f10d6446c6ac478d Mon Sep 17 00:00:00 2001
From: Jerry D. Hedden <jdhedden@cpan.org>
Date: Fri, 24 Sep 2010 14:45:29 -0400
Subject: [PATCH] Upgrade to threads 1.79

---
 MANIFEST                    |    1 +
 Porting/Maintainers.pl      |    2 +-
 dist/threads/lib/threads.pm |   12 ++++-
 dist/threads/t/exit.t       |   10 ++--
 dist/threads/t/kill2.t      |   91 +++++++++++++++++++++++++++++++++++++++++++
 dist/threads/t/thread.t     |    2 +-
 dist/threads/threads.xs     |   48 +++++++++++++++--------
 7 files changed, 140 insertions(+), 26 deletions(-)
 create mode 100644 dist/threads/t/kill2.t

diff --git a/MANIFEST b/MANIFEST
index c935b4e..9a20b90 100644
--- a/MANIFEST
+++ b/MANIFEST
@@ -2984,6 +2984,7 @@ dist/threads/t/free2.t		More ithread destruction tests
 dist/threads/t/free.t		Test ithread destruction
 dist/threads/threads.xs		ithreads
 dist/threads/t/join.t		Testing the join function
+dist/threads/t/kill2.t		Tests thread signalling
 dist/threads/t/kill.t		Tests thread signalling
 dist/threads/t/libc.t		testing libc functions for threadsafety
 dist/threads/t/list.t		Test threads->list()
diff --git a/Porting/Maintainers.pl b/Porting/Maintainers.pl
index fb65347..46171f3 100755
--- a/Porting/Maintainers.pl
+++ b/Porting/Maintainers.pl
@@ -1436,7 +1436,7 @@ use File::Glob qw(:case);
     'threads' =>
 	{
 	'MAINTAINER'	=> 'jdhedden',
-	'DISTRIBUTION'	=> 'JDHEDDEN/threads-1.77.tar.gz',
+	'DISTRIBUTION'	=> 'JDHEDDEN/threads-1.79.tar.gz',
 	'FILES'		=> q[dist/threads],
 	'EXCLUDED'	=> [ qr{^examples/},
 			     qw(t/pod.t
diff --git a/dist/threads/lib/threads.pm b/dist/threads/lib/threads.pm
index 8654f9f..e5ceb54 100644
--- a/dist/threads/lib/threads.pm
+++ b/dist/threads/lib/threads.pm
@@ -5,7 +5,7 @@ use 5.008;
 use strict;
 use warnings;
 
-our $VERSION = '1.77_03';
+our $VERSION = '1.79_01';
 my $XS_VERSION = $VERSION;
 $VERSION = eval $VERSION;
 
@@ -134,7 +134,7 @@ threads - Perl interpreter-based threads
 
 =head1 VERSION
 
-This document describes threads version 1.77
+This document describes threads version 1.79
 
 =head1 SYNOPSIS
 
@@ -1005,6 +1005,12 @@ mutexes that are needed to control functionality within the L<threads> module.
 For this reason, the use of C<END> blocks in threads is B<strongly>
 discouraged.
 
+=item Open directory handles
+
+Spawning threads with open directory handles (see
+L<opendir|perlfunc/"opendir DIRHANDLE,EXPR">) will crash the interpreter.
+L<[perl #75154]|http://rt.perl.org/rt3/Public/Bug/Display.html?id=75154>
+
 =item Perl Bugs and the CPAN Version of L<threads>
 
 Support for threads extends beyond the code in this module (i.e.,
@@ -1034,7 +1040,7 @@ L<threads> Discussion Forum on CPAN:
 L<http://www.cpanforum.com/dist/threads>
 
 Annotated POD for L<threads>:
-L<http://annocpan.org/~JDHEDDEN/threads-1.77/threads.pm>
+L<http://annocpan.org/~JDHEDDEN/threads-1.79/threads.pm>
 
 Source repository:
 L<http://code.google.com/p/threads-shared/>
diff --git a/dist/threads/t/exit.t b/dist/threads/t/exit.t
index 29c3dca..208f7b3 100644
--- a/dist/threads/t/exit.t
+++ b/dist/threads/t/exit.t
@@ -48,7 +48,7 @@ my $rc = $thr->join();
 ok(! defined($rc), 'Exited: threads->exit()');
 
 
-run_perl(prog => 'use threads 1.77;' .
+run_perl(prog => 'use threads 1.79;' .
                  'threads->exit(86);' .
                  'exit(99);',
          nolib => ($ENV{PERL_CORE}) ? 0 : 1,
@@ -98,7 +98,7 @@ $rc = $thr->join();
 ok(! defined($rc), 'Exited: $thr->set_thread_exit_only');
 
 
-run_perl(prog => 'use threads 1.77 qw(exit thread_only);' .
+run_perl(prog => 'use threads 1.79 qw(exit thread_only);' .
                  'threads->create(sub { exit(99); })->join();' .
                  'exit(86);',
          nolib => ($ENV{PERL_CORE}) ? 0 : 1,
@@ -108,7 +108,7 @@ run_perl(prog => 'use threads 1.77 qw(exit thread_only);' .
     is($?>>8, 86, "'use threads 'exit' => 'thread_only'");
 }
 
-my $out = run_perl(prog => 'use threads 1.77;' .
+my $out = run_perl(prog => 'use threads 1.79;' .
                            'threads->create(sub {' .
                            '    exit(99);' .
                            '});' .
@@ -124,7 +124,7 @@ my $out = run_perl(prog => 'use threads 1.77;' .
 like($out, '1 finished and unjoined', "exit(status) in thread");
 
 
-$out = run_perl(prog => 'use threads 1.77 qw(exit thread_only);' .
+$out = run_perl(prog => 'use threads 1.79 qw(exit thread_only);' .
                         'threads->create(sub {' .
                         '   threads->set_thread_exit_only(0);' .
                         '   exit(99);' .
@@ -141,7 +141,7 @@ $out = run_perl(prog => 'use threads 1.77 qw(exit thread_only);' .
 like($out, '1 finished and unjoined', "set_thread_exit_only(0)");
 
 
-run_perl(prog => 'use threads 1.77;' .
+run_perl(prog => 'use threads 1.79;' .
                  'threads->create(sub {' .
                  '   $SIG{__WARN__} = sub { exit(99); };' .
                  '   die();' .
diff --git a/dist/threads/t/kill2.t b/dist/threads/t/kill2.t
new file mode 100644
index 0000000..8eac55b
--- /dev/null
+++ b/dist/threads/t/kill2.t
@@ -0,0 +1,91 @@
+use strict;
+use warnings;
+
+BEGIN {
+    require($ENV{PERL_CORE} ? '../../t/test.pl' : './t/test.pl');
+
+    use Config;
+    if (! $Config{'useithreads'}) {
+        skip_all(q/Perl not compiled with 'useithreads'/);
+    }
+}
+
+use ExtUtils::testlib;
+
+use threads;
+
+BEGIN {
+    if (! eval 'use threads::shared; 1') {
+        skip_all('threads::shared not available');
+    }
+
+    local $SIG{'HUP'} = sub {};
+    my $thr = threads->create(sub {});
+    eval { $thr->kill('HUP') };
+    $thr->join();
+    if ($@ && $@ =~ /safe signals/) {
+        skip_all('Not using safe signals');
+    }
+
+    plan(3);
+};
+
+fresh_perl_is(<<'EOI', 'ok', { }, 'No signal handler in thread');
+    use threads;
+    use Thread::Semaphore;
+    my $sema = Thread::Semaphore->new(0);
+    my $test = sub {
+        my $sema = shift;
+        $sema->up();
+        while(1) { sleep(1); }
+    };
+    my $thr = threads->create($test, $sema);
+    $sema->down();
+    $thr->detach();
+    eval {
+        $thr->kill('STOP');
+    };
+    print(($@ =~ /no signal handler set/) ? 'ok' : 'not ok');
+EOI
+
+fresh_perl_is(<<'EOI', 'ok', { }, 'Handler to signal mismatch');
+    use threads;
+    use Thread::Semaphore;
+    my $sema = Thread::Semaphore->new(0);
+    my $test = sub {
+        my $sema = shift;
+        $SIG{'TERM'} = sub { threads->exit() };
+        $sema->up();
+        while(1) { sleep(1); }
+    };
+    my $thr = threads->create($test, $sema);
+    $sema->down();
+    $thr->detach();
+    eval {
+        $thr->kill('STOP');
+    };
+    print(($@ =~ /no signal handler set/) ? 'ok' : 'not ok');
+EOI
+
+fresh_perl_is(<<'EOI', 'ok', { }, 'Handler and signal match');
+    use threads;
+    use Thread::Semaphore;
+    my $sema = Thread::Semaphore->new(0);
+    my $test = sub {
+        my $sema = shift;
+        $SIG{'STOP'} = sub { threads->exit() };
+        $sema->up();
+        while(1) { sleep(1); }
+    };
+    my $thr = threads->create($test, $sema);
+    $sema->down();
+    $thr->detach();
+    eval {
+        $thr->kill('STOP');
+    };
+    print((! $@) ? 'ok' : 'not ok');
+EOI
+
+exit(0);
+
+# EOF
diff --git a/dist/threads/t/thread.t b/dist/threads/t/thread.t
index b390215..32c50b8 100644
--- a/dist/threads/t/thread.t
+++ b/dist/threads/t/thread.t
@@ -161,7 +161,7 @@ package main;
 
 # bugid #24165
 
-run_perl(prog => 'use threads 1.77;' .
+run_perl(prog => 'use threads 1.79;' .
                  'sub a{threads->create(shift)} $t = a sub{};' .
                  '$t->tid; $t->join; $t->tid',
          nolib => ($ENV{PERL_CORE}) ? 0 : 1,
diff --git a/dist/threads/threads.xs b/dist/threads/threads.xs
index f4e6946..d5e20b9 100644
--- a/dist/threads/threads.xs
+++ b/dist/threads/threads.xs
@@ -1342,6 +1342,7 @@ ithread_kill(...)
         ithread *thread;
         char *sig_name;
         IV signal;
+        int no_handler = 1;
     CODE:
         /* Must have safe signals */
         if (PL_signals & PERL_SIGNALS_UNSAFE_FLAG) {
@@ -1371,11 +1372,21 @@ ithread_kill(...)
         MUTEX_LOCK(&thread->mutex);
         if (thread->interp) {
             dTHXa(thread->interp);
-            PL_psig_pend[signal]++;
-            PL_sig_pending = 1;
+            if (PL_psig_pend && PL_psig_ptr[signal]) {
+                PL_psig_pend[signal]++;
+                PL_sig_pending = 1;
+                no_handler = 0;
+            }
+        } else {
+            /* Ignore signal to terminated thread */
+            no_handler = 0;
         }
         MUTEX_UNLOCK(&thread->mutex);
 
+        if (no_handler) {
+            Perl_croak(aTHX_ "Signal %s received in thread %"UVuf", but no signal handler set.", sig_name, thread->tid);
+        }
+
         /* Return the thread to allow for method chaining */
         ST(0) = ST(0);
         /* XSRETURN(1); - implied */
-- 
1.6.1.2


```
</details>

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @bingos

On Fri\, Sep 24\, 2010 at 03&#8203;:11&#8203;:40PM \-0400\, Jerry D\. Hedden wrote&#8203;:
> Attached is a revised patch to upgrade to v1\.79\.  I needed to fix
> the tests in the newly added t/kill2\.t file\.  The blead version will
> be 1\.79\_01\.
> 
> On Wed\, Sep 22\, 2010 at 16&#8203;:08\, Jerry D\. Hedden
> \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> > \# New Ticket Created by ?"Jerry D\. Hedden"
> > \# Please include the string&#8203;: ?\[perl \#78000\]
> > \# in the subject line of all future correspondence about this issue\.
> > \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=78000 >
> >
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > The attached patch upgrades the 'threads' module to v1\.78\.
> >
> > This upgrade fixes 'threads' bug \#60460
> > ? ?https://rt.cpan.org/Ticket/Display.html?id=60460
> > which results in segfaults if a signal is sent to a thread\, and
> > there is no corresponding signal handler set in the thread\.
> > Corresponding tests are included\.
> >
> > Additionally\, the POD is updated to document \[perl \#75154\] concerning
> > open dir handles and threads\.
> >
> > The module version for blead is set to 1\.78\_01 to account for
> > blead changes not present in the CPAN version\.
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> 
> 

Thanks\, applied to blead as

http&#8203;://perl5\.git\.perl\.org/perl\.git/commitdiff/414fa04cdda19d4646d695b14dd909e515b70524

\-\- 
Chris Williams
aka BinGOs
PGP ID 0x4658671F
http&#8203;://www\.gumbynet\.org\.uk
==========================

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
